### DISCUSSION TITLE
Stricter typings for `on` and `emit` on client and server sides

### DISCUSSION THREAD

**Discussion by david-fong** [socketio/socket.io#3777] [Ideas] [answered]
Here is a [TypeScript Playground link](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBDAnmApnA3nAyhAxgaxXgEMBnOAITJQGEAbYFAOxhwKLgF84AzKCEHABEpPIRgA6YBAC0uBsxhCA3AFgAUKEiwEyNJjbi4ZStSwooANwuGO3PgOGj2k6So0bgLCz2K40NAIgpAAiKKS4GBoAkFAoxAAmEEx0iHAAqmCkMHHEIABccEUASii40AkAPNlQXgDmADRwuUkpacRMiADaALoAfGrqsfGtqXAhEADuTDXxBXCl5VBVNfVNLcljHd39g5we6gD0AFTHcMeHGoeHcAACMKQyKAAeqLgwT1D8UBov2vBeGA+PwBBQsLA5YDvWwwSo0OAvIFMBLkQIgYJhCJ9BHPJEo0ykWhg1hiDjoA7RE5nW4QaxfYAJNAXGLJSoAURxePIhEQEB4cBoXSEE2mszyQn6AAoUNYWAA5PIoQpspq4Yh0OgAIz8+EKkokBuIUDqpEKguFUxmOTmEq6bJ6AEo4ABebGWCAMh2FGAAC2ApEGMSpd1pFlqjPOlyGKBAwFhHMRzHxPL5AqFmTFIAlfWlspgCpASrgKrgBokRpNZvTWWt4p6dsd3r9AY0+3UV1OkauN3ujz+ZQ+Yegv1e0AB3igvn82AsdIhtWhpNh8MTyNRQVC4Vw2NX+KohPMVhsS6i7aGwZpdPDTKj0VZCdxSe5KF5-PNGdrWalMsUBaLJbVDVtQIPUywrU00yED9ci-BsnVdOB3U9Jt-UDc9O0vMMGRvGIYzjdlOSfOAUzfIURStGDs1zX9FWVJowONCDzXIzNbXtL04E1CAIDoeImD2A5uxdZ0RNEsTxIkySpOkmTZIkoS2WePIwF400Di0McMDgaRjFRYkAEkIC4Xh+EEEQlykWR5EYFh3E0cBNIMWcLF0mcjygJoYVcw86X3FAvPsUynAstw0L+TSmEVUgwBBcYoRgaRItqcJT2icKdEBYFpxCABGVLhkSTY0mguZCnJIZomiIQEhymQAFcaxgmQAEEhEKLomDqkBNQsJpViYRoMGIQpOu6ixOB6QZKqqmr6sauYZAoNq4C6frBvQYa4FGnqoAmqbKuq2qGszGQaGWroNpGrqdqaTVCjW1VCgAMTqph3kSvaYmiTh9o2NpxktTMyq+mbaqSUVPxkHLzu23qhqusbdqaCBNQAKwHejDU6XpJpBw6ZHBiiFoAJnOtbcYq0GCcByGAGYYeuuHYY8rbGZZh64AANQ9BIXrehLkgpyqfpiNsiiKa44AAWWgNAEnixKjUYcgTCYFAUEZBIy1bA5yitOBrMUQw8udAUDIgSV0E4J0THoGzWEhRcXEqEIFeSJXwgkXKBg0Q3wTEHKJGSSV8cJk7oaaSViAABiaYgcrj4m49puOABY44AVng7FyuiePyz2B1Bj9kkCEDvCYBD2bjshpamiEIQmkuuBiettC1cmNy6Ut62g54EOZgjuBJW1A8xEKHyLD8mFs9SvXsjgQl3ONl1OLMMRvOcqB53imEXbdpLla9nKfaGJe5wDoOmCro75ryFrG+HmO44T4xidn3PoimtuYnP48y4kBXG+1MIZNXpk0V+Sc4Ap2EI-SUs9XqMh4F4DWRdWxFyAA) for an idea I've been thinking about for a while.

[Here is a link to the project repo](https://github.com/david-fong/socket.io-event-types).

It allows to specify explicit typings / function signatures for `on` and `emit`. This could be used to _enable_ the compiler to warn when the arguments are not being passed properly. I like the "Upstream" / "Downstream" dictionary interface. Try it out for yourself and share what you think. Is this something you think you could benefit from?

The next thing I will add is an optional last argument for an ack function.

**Comment by darrachequesne** [comment, marked as answer]
For future readers:

This was implemented in https://github.com/socketio/socket.io/commit/0107510ba8a0f148c78029d8be8919b350feb633, included in version [4.0.0](https://github.com/socketio/socket.io/releases/tag/4.0.0) :rocket: 

Documentation: https://socket.io/docs/v4/typescript/
